### ISSUE TITLE
ICE On unknown type on HapMap templated

### ISSUE THREAD

**Issue by AurelienFT** [rust-lang/rust#98620]
### Code

```Rust
use std::collections::HashMap;
use std::hash::BuildHasherDefault;

pub type BuildMap<T> = BuildHasherDefault<T>;

pub type Map<K, V> = HashMap<K, V, BuildMap<K>>;

fn main() {
    let length = 3;
    let mut new_operations = Map::with_capacity_and_hasher(length, BuildMap::default());
    if !new_operations.is_empty() {
        return;
    }
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.63.0-nightly (5435ed691 2022-06-07)
binary: rustc
commit-hash: 5435ed6916a59e8d5acba2149316a841c3905cbd
commit-date: 2022-06-07
host: aarch64-apple-darwin
release: 1.63.0-nightly
LLVM version: 14.0.5
```

### Error output

```
error: internal compiler error: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:856:33: unexpected path: def=std::collections::HashMap substs=[_, _, _] path=Path { span: src/main.rs:22:30: 22:33 (#0), res: Def(TyAlias, DefId(0:20 ~ tmp[b4a0]::Map)), segments: [PathSegment { ident: Map#0, hir_id: Some(HirId { owner: DefId(0:23 ~ tmp[b4a0]::main), local_id: 6 }), res: Some(Err), args: None, infer_args: true }] }
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'Box<dyn Any>', /rustc/5435ed6916a59e8d5acba2149316a841c3905cbd/compiler/rustc_errors/src/lib.rs:1334:9
stack backtrace:
   0:        0x103a44090 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h289ddc002351b523
   1:        0x103a95a6c - core::fmt::write::h799deef4ed1d694b
   2:        0x103a3718c - std::io::Write::write_fmt::h3e628554da7ceb9c
   3:        0x103a46cd4 - std::panicking::default_hook::{{closure}}::ha83375eb6d56849c
   4:        0x103a46a3c - std::panicking::default_hook::h8ca7f4d98f9cfccf
   5:        0x10ae703cc - rustc_driver[6012721eacdbdd6d]::DEFAULT_HOOK::{closure#0}::{closure#0}
   6:        0x103a472a4 - std::panicking::rust_panic_with_hook::h69563d1f3c0e4475
   7:        0x10e958270 - std[3f1a734e0e85718c]::panicking::begin_panic::<rustc_errors[5dd2c2aa9e856d8b]::ExplicitBug>::{closure#0}
   8:        0x10e956548 - std[3f1a734e0e85718c]::sys_common::backtrace::__rust_end_short_backtrace::<std[3f1a734e0e85718c]::panicking::begin_panic<rustc_errors[5dd2c2aa9e856d8b]::ExplicitBug>::{closure#0}, !>
   9:        0x10ed8886c - std[3f1a734e0e85718c]::panicking::begin_panic::<rustc_errors[5dd2c2aa9e856d8b]::ExplicitBug>
  10:        0x10e8d480c - std[3f1a734e0e85718c]::panic::panic_any::<rustc_errors[5dd2c2aa9e856d8b]::ExplicitBug>
  11:        0x10e8d455c - <rustc_errors[5dd2c2aa9e856d8b]::HandlerInner>::bug::<&alloc[a2584a6b23f7fae9]::string::String>
  12:        0x10e8d2ae4 - <rustc_errors[5dd2c2aa9e856d8b]::Handler>::bug::<&alloc[a2584a6b23f7fae9]::string::String>
  13:        0x10e961efc - rustc_middle[f73a57d5dc2f596d]::ty::context::tls::with_opt::<rustc_middle[f73a57d5dc2f596d]::util::bug::opt_span_bug_fmt<rustc_span[3a27d002c0bb3dbf]::span_encoding::Span>::{closure#0}, ()>
  14:        0x10e961f70 - rustc_middle[f73a57d5dc2f596d]::util::bug::opt_span_bug_fmt::<rustc_span[3a27d002c0bb3dbf]::span_encoding::Span>
  15:        0x10ed8d7d4 - rustc_middle[f73a57d5dc2f596d]::util::bug::bug_fmt
  16:        0x10e79e97c - <rustc_infer[be8276cf2602c983]::infer::error_reporting::need_type_info::FindInferSourceVisitor>::path_inferred_subst_iter
  17:        0x10e79f00c - <rustc_infer[be8276cf2602c983]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[ff83a92110a313d3]::intravisit::Visitor>::visit_expr
  18:        0x10e79edf4 - <rustc_infer[be8276cf2602c983]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[ff83a92110a313d3]::intravisit::Visitor>::visit_expr
  19:        0x10e79e9d8 - <rustc_infer[be8276cf2602c983]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[ff83a92110a313d3]::intravisit::Visitor>::visit_local
  20:        0x10e7bafac - rustc_hir[ff83a92110a313d3]::intravisit::walk_expr::<rustc_infer[be8276cf2602c983]::infer::error_reporting::need_type_info::FindInferSourceVisitor>
  21:        0x10e79ef1c - <rustc_infer[be8276cf2602c983]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[ff83a92110a313d3]::intravisit::Visitor>::visit_expr
  22:        0x10e73e80c - <rustc_infer[be8276cf2602c983]::infer::InferCtxt>::emit_inference_failure_err
  23:        0x10e63eddc - <rustc_infer[be8276cf2602c983]::infer::InferCtxt as rustc_trait_selection[85ea576f6cb7c13d]::traits::error_reporting::InferCtxtPrivExt>::maybe_report_ambiguity
  24:        0x10e63213c - <rustc_infer[be8276cf2602c983]::infer::InferCtxt as rustc_trait_selection[85ea576f6cb7c13d]::traits::error_reporting::InferCtxtExt>::report_fulfillment_errors
  25:        0x10d28815c - <rustc_typeck[4f19e3822f8d5d40]::check::fn_ctxt::FnCtxt>::select_all_obligations_or_error
  26:        0x10d383c7c - <rustc_infer[be8276cf2602c983]::infer::InferCtxtBuilder>::enter::<&rustc_middle[f73a57d5dc2f596d]::ty::context::TypeckResults, <rustc_typeck[4f19e3822f8d5d40]::check::inherited::InheritedBuilder>::enter<rustc_typeck[4f19e3822f8d5d40]::check::typeck_with_fallback<rustc_typeck[4f19e3822f8d5d40]::check::typeck::{closure#0}>::{closure#1}, &rustc_middle[f73a57d5dc2f596d]::ty::context::TypeckResults>::{closure#0}>
  27:        0x10d2fd464 - rustc_typeck[4f19e3822f8d5d40]::check::typeck
  28:        0x10de61808 - <rustc_query_system[136bb7c139dddbdd]::dep_graph::graph::DepGraph<rustc_middle[f73a57d5dc2f596d]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[f73a57d5dc2f596d]::ty::context::TyCtxt, rustc_span[3a27d002c0bb3dbf]::def_id::LocalDefId, &rustc_middle[f73a57d5dc2f596d]::ty::context::TypeckResults>
  29:        0x10dc0a3b8 - rustc_query_system[136bb7c139dddbdd]::query::plumbing::try_execute_query::<rustc_query_impl[2e8a5a91a394c2ab]::plumbing::QueryCtxt, rustc_query_system[136bb7c139dddbdd]::query::caches::DefaultCache<rustc_span[3a27d002c0bb3dbf]::def_id::LocalDefId, &rustc_middle[f73a57d5dc2f596d]::ty::context::TypeckResults>>
  30:        0x10dce218c - rustc_query_system[136bb7c139dddbdd]::query::plumbing::get_query::<rustc_query_impl[2e8a5a91a394c2ab]::queries::typeck, rustc_query_impl[2e8a5a91a394c2ab]::plumbing::QueryCtxt>
  31:        0x10d427ca8 - <rustc_middle[f73a57d5dc2f596d]::hir::map::Map>::par_body_owners::<rustc_typeck[4f19e3822f8d5d40]::check::typeck_item_bodies::{closure#0}>
  32:        0x10d301508 - rustc_typeck[4f19e3822f8d5d40]::check::typeck_item_bodies
  33:        0x10de8a154 - <rustc_query_system[136bb7c139dddbdd]::dep_graph::graph::DepGraph<rustc_middle[f73a57d5dc2f596d]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[f73a57d5dc2f596d]::ty::context::TyCtxt, (), ()>
  34:        0x10dc4c9ac - rustc_query_system[136bb7c139dddbdd]::query::plumbing::try_execute_query::<rustc_query_impl[2e8a5a91a394c2ab]::plumbing::QueryCtxt, rustc_query_system[136bb7c139dddbdd]::query::caches::DefaultCache<(), ()>>
  35:        0x10dcb142c - rustc_query_system[136bb7c139dddbdd]::query::plumbing::get_query::<rustc_query_impl[2e8a5a91a394c2ab]::queries::typeck_item_bodies, rustc_query_impl[2e8a5a91a394c2ab]::plumbing::QueryCtxt>
  36:        0x10d3b0b14 - <rustc_session[4316a371e6e67229]::session::Session>::time::<(), rustc_typeck[4f19e3822f8d5d40]::check_crate::{closure#7}>
  37:        0x10d3789b4 - rustc_typeck[4f19e3822f8d5d40]::check_crate
  38:        0x10af75574 - rustc_interface[b81c2f735902673e]::passes::analysis
  39:        0x10de81ddc - <rustc_query_system[136bb7c139dddbdd]::dep_graph::graph::DepGraph<rustc_middle[f73a57d5dc2f596d]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[f73a57d5dc2f596d]::ty::context::TyCtxt, (), core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>>
  40:        0x10dc41654 - rustc_query_system[136bb7c139dddbdd]::query::plumbing::try_execute_query::<rustc_query_impl[2e8a5a91a394c2ab]::plumbing::QueryCtxt, rustc_query_system[136bb7c139dddbdd]::query::caches::DefaultCache<(), core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>>>
  41:        0x10dce25e0 - rustc_query_system[136bb7c139dddbdd]::query::plumbing::get_query::<rustc_query_impl[2e8a5a91a394c2ab]::queries::analysis, rustc_query_impl[2e8a5a91a394c2ab]::plumbing::QueryCtxt>
  42:        0x10ae33ecc - <rustc_interface[b81c2f735902673e]::passes::QueryContext>::enter::<rustc_driver[6012721eacdbdd6d]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>>
  43:        0x10ae16fd4 - <rustc_interface[b81c2f735902673e]::interface::Compiler>::enter::<rustc_driver[6012721eacdbdd6d]::run_compiler::{closure#1}::{closure#2}, core[aa3fa2e4c8a303d9]::result::Result<core[aa3fa2e4c8a303d9]::option::Option<rustc_interface[b81c2f735902673e]::queries::Linker>, rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>>
  44:        0x10ae036d8 - rustc_span[3a27d002c0bb3dbf]::with_source_map::<core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>, rustc_interface[b81c2f735902673e]::interface::create_compiler_and_run<core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>, rustc_driver[6012721eacdbdd6d]::run_compiler::{closure#1}>::{closure#1}>
  45:        0x10ae177c0 - <scoped_tls[a5215fda6f8a877a]::ScopedKey<rustc_span[3a27d002c0bb3dbf]::SessionGlobals>>::set::<rustc_interface[b81c2f735902673e]::interface::run_compiler<core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>, rustc_driver[6012721eacdbdd6d]::run_compiler::{closure#1}>::{closure#0}, core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>>
  46:        0x10ae3640c - std[3f1a734e0e85718c]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[b81c2f735902673e]::util::run_in_thread_pool_with_globals<rustc_interface[b81c2f735902673e]::interface::run_compiler<core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>, rustc_driver[6012721eacdbdd6d]::run_compiler::{closure#1}>::{closure#0}, core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>>::{closure#0}, core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>>
  47:        0x10ae36c84 - <<std[3f1a734e0e85718c]::thread::Builder>::spawn_unchecked_<rustc_interface[b81c2f735902673e]::util::run_in_thread_pool_with_globals<rustc_interface[b81c2f735902673e]::interface::run_compiler<core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>, rustc_driver[6012721eacdbdd6d]::run_compiler::{closure#1}>::{closure#0}, core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>>::{closure#0}, core[aa3fa2e4c8a303d9]::result::Result<(), rustc_errors[5dd2c2aa9e856d8b]::ErrorGuaranteed>>::{closure#1} as core[aa3fa2e4c8a303d9]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  48:        0x103a4fb00 - std::sys::unix::thread::Thread::new::thread_start::h65074f5cfbbe5af3
  49:        0x18ccd026c - __pthread_deallocate

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.63.0-nightly (5435ed691 2022-06-07) running on aarch64-apple-darwin

note: compiler flags: --crate-type bin -C embed-bitcode=no -C split-debuginfo=unpacked -C debuginfo=2 -C incremental

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [typeck] type-checking `main`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `tmp`
```

</p>
</details>

**Comment by Rageking8** [comment]
Seems to have already been fixed with the latest Nightly version: 1.64.0-nightly (2022-06-27 2f3ddd9f594adf977354). Could not reproduce ICE using rust playground for latest nightly, beta and stable 1.61.

**Comment by AurelienFT** [comment]
Yes it has been fixed. Just test it I close it.
